File: packages/sdk-go/pkg/types/canary.go
package types

import (
	"time"

	"github.com/google/uuid"
)

// CanaryRolloutState enumerates the states of a canary rollout lifecycle.
//
// Transitions (enforced by reconciler/canary.go):
//
//	pending → running → validating → promoting → succeeded
//	                  ↘ auto_rolled_back
//	                  ↘ manual_rolled_back
//	    ↘ failed (from any non-terminal state on reconciler error)
//
// Terminal states (no further transitions): succeeded, auto_rolled_back,
// manual_rolled_back, failed.
type CanaryRolloutState string

const (
	CanaryStatePending          CanaryRolloutState = "pending"
	CanaryStateRunning          CanaryRolloutState = "running"
	CanaryStateValidating       CanaryRolloutState = "validating"
	CanaryStatePromoting        CanaryRolloutState = "promoting"
	CanaryStateSucceeded        CanaryRolloutState = "succeeded"
	CanaryStateAutoRolledBack   CanaryRolloutState = "auto_rolled_back"
	CanaryStateManualRolledBack CanaryRolloutState = "manual_rolled_back"
	CanaryStateFailed           CanaryRolloutState = "failed"
)

// IsTerminal reports whether the rollout has reached an end state.
func (s CanaryRolloutState) IsTerminal() bool {
	switch s {
	case CanaryStateSucceeded, CanaryStateAutoRolledBack, CanaryStateManualRolledBack, CanaryStateFailed:
		return true
	}
	return false
}

// IsActive reports whether the rollout is currently in flight (non-terminal).
func (s CanaryRolloutState) IsActive() bool {
	return !s.IsTerminal()
}

// CanaryRollout represents an in-flight or historical canary release.
//
// The two-Deployment pattern is realized in K8s as:
//   - <service>         — the existing stable Deployment (continues to serve traffic)
//   - <service>-canary  — a new Deployment spun up at `CanaryReplicas` count,
//     selected by the same Service via the shared `app=<svc>` label.
//
// Traffic split is purely replica-count proportion (no service mesh). For a
// 20% canary at 5 total replicas: 4 stable + 1 canary.
//
// On auto-promote, a new Deployment `<service>-stable-new` is created from
// the canary digest, scaled up, health-gated, then the old stable is scaled
// to 0 and deleted after a 15-minute soak — see Reconciler.Promote.
type CanaryRollout struct {
	ID            uuid.UUID `json:"id" db:"id"`
	ServiceID     uuid.UUID `json:"service_id" db:"service_id"`
	EnvironmentID uuid.UUID `json:"environment_id" db:"environment_id"`

	StableDeploymentID    uuid.UUID  `json:"stable_deployment_id" db:"stable_deployment_id"`
	CanaryDeploymentID    uuid.UUID  `json:"canary_deployment_id" db:"canary_deployment_id"`
	NewStableDeploymentID *uuid.UUID `json:"new_stable_deployment_id,omitempty" db:"new_stable_deployment_id"`

	// Rollout spec (frozen at start)
	CanaryDigest            string  `json:"canary_digest" db:"canary_digest"`
	CanaryPercentage        int     `json:"canary_percentage" db:"canary_percentage"`
	TotalReplicas           int     `json:"total_replicas" db:"total_replicas"`
	CanaryReplicas          int     `json:"canary_replicas" db:"canary_replicas"`
	StableReplicas          int     `json:"stable_replicas" db:"stable_replicas"`
	ValidationWindowSeconds int     `json:"validation_window_seconds" db:"validation_window_seconds"`
	SmokeEndpoint           string  `json:"smoke_endpoint,omitempty" db:"smoke_endpoint"`
	ErrorRateThreshold      float64 `json:"error_rate_threshold" db:"error_rate_threshold"`

	// State machine
	State               CanaryRolloutState `json:"state" db:"state"`
	StartedAt           *time.Time         `json:"started_at,omitempty" db:"started_at"`
	ValidatingStartedAt *time.Time         `json:"validating_started_at,omitempty" db:"validating_started_at"`
	PromotingStartedAt  *time.Time         `json:"promoting_started_at,omitempty" db:"promoting_started_at"`
	TerminalAt          *time.Time         `json:"terminal_at,omitempty" db:"terminal_at"`

	// Audit / diagnostics
	InitiatedBy     *uuid.UUID `json:"initiated_by,omitempty" db:"initiated_by"`
	ChangeTicketURL string     `json:"change_ticket_url,omitempty" db:"change_ticket_url"`
	LastError       string     `json:"last_error,omitempty" db:"last_error"`
	RollbackReason  string     `json:"rollback_reason,omitempty" db:"rollback_reason"`

	CreatedAt time.Time `json:"created_at" db:"created_at"`
	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
}

// CanaryRolloutSpec is the user-supplied configuration for a rollout.
// Validation lives on this struct so it can be reused by API + CLI.
type CanaryRolloutSpec struct {
	// ImageDigest of the candidate (must already be built — reuse an existing
	// Release). The caller is responsible for resolving a service.yaml deploy
	// into a concrete digest before calling the API.
	ImageDigest string `json:"digest"`
	// Percentage of traffic to route to the canary (5-50).
	Percentage int `json:"percentage"`
	// ValidationWindowMinutes is how long the canary must stay healthy before
	// auto-promotion triggers. Range: 1-60.
	ValidationWindowMinutes int `json:"validation_window_minutes"`
	// SmokeEndpoint (optional) is a URL Path on the canary Service (e.g.
	// "/health/deep") that must return 200 during the validation window.
	SmokeEndpoint string `json:"smoke_endpoint,omitempty"`
	// ErrorRateThreshold allows at most this fraction of 5xx responses during
	// validation. Defaults to 0.05 (5%). Range: 0.0-0.5.
	ErrorRateThreshold float64 `json:"error_rate_threshold,omitempty"`
	// ChangeTicketURL is required for production environments (matches the
	// DeployService / InstantRollback HITL gate).
	ChangeTicketURL string `json:"change_ticket_url,omitempty"`
}

File: packages/sdk-py/src/enclii_sdk/models/canary.py
"""Canary rollout models (P2.7).

Replica-proportion traffic splitting with auto-promote after a
validation window. See ``apps/switchyard-api/internal/reconciler/canary.go``
for the state machine authority.
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from enclii_sdk.models.common import StrEnum


class CanaryRolloutState(StrEnum):
    """State machine for a canary rollout.

    Transitions::

        pending → running → validating → promoting → succeeded
                          ↘ auto_rolled_back
                          ↘ manual_rolled_back
            ↘ failed (from any non-terminal on reconciler error)
    """

    PENDING = "pending"
    RUNNING = "running"
    VALIDATING = "validating"
    PROMOTING = "promoting"
    SUCCEEDED = "succeeded"
    AUTO_ROLLED_BACK = "auto_rolled_back"
    MANUAL_ROLLED_BACK = "manual_rolled_back"
    FAILED = "failed"

    def is_terminal(self) -> bool:
        """Return ``True`` if the rollout is in an end state."""
        return self in (
            CanaryRolloutState.SUCCEEDED,
            CanaryRolloutState.AUTO_ROLLED_BACK,
            CanaryRolloutState.MANUAL_ROLLED_BACK,
            CanaryRolloutState.FAILED,
        )

    def is_active(self) -> bool:
        """Return ``True`` if the rollout is still in flight."""
        return not self.is_terminal()


class StartCanaryRequest(BaseModel):
    """Payload for ``POST /v1/services/{id}/canary``.

    Validation bounds (enforced server-side by the reconciler):
      - ``percentage``: 5-50
      - ``validation_window_minutes``: 1-60
      - ``error_rate_threshold``: 0.0-0.5 (default 0.05)
    """

    digest: str = Field(description="Image digest or Release UUID of the candidate")
    percentage: int = Field(ge=5, le=50)
    validation_window_minutes: int = Field(default=10, ge=1, le=60)
    smoke_endpoint: str = ""
    error_rate_threshold: float = Field(default=0.05, ge=0.0, le=0.5)
    environment_name: str = "production"
    change_ticket_url: str = ""
    total_replicas: int = 0


class CanaryRollout(BaseModel):
    """A single in-flight or historical canary rollout."""

    model_config = ConfigDict(extra="allow")

    id: UUID
    service_id: UUID
    environment_id: UUID
    stable_deployment_id: UUID
    canary_deployment_id: UUID
    new_stable_deployment_id: UUID | None = None
    canary_digest: str
    canary_percentage: int
    total_replicas: int
    canary_replicas: int
    stable_replicas: int
    validation_window_seconds: int
    smoke_endpoint: str = ""
    error_rate_threshold: float = 0.05
    state: CanaryRolloutState
    started_at: datetime | None = None
    validating_started_at: datetime | None = None
    promoting_started_at: datetime | None = None
    terminal_at: datetime | None = None
    initiated_by: UUID | None = None
    change_ticket_url: str = ""
    last_error: str = ""
    rollback_reason: str = ""
    created_at: datetime
    updated_at: datetime


class CanaryRolloutResponse(CanaryRollout):
    """Canary rollout with effective traffic share after replica rounding."""

    actual_percentage: float


__all__ = [
    "CanaryRollout",
    "CanaryRolloutResponse",
    "CanaryRolloutState",
    "StartCanaryRequest",
]
